astropy/wcs/utils.py

```diff
@@ -635,193 +585,225 @@ def pixel_to_skycoord(xp, yp, wcs, origin=0, mode='all', cls=None):
-            else:
-                pixel_inputs.append(inputs[ipix].flat[0])
-
-        pixel_inputs = np.broadcast_arrays(*pixel_inputs)
-
-        world_outputs = wcs_in.pixel_to_world(*pixel_inputs)
-
-        if not isinstance(world_outputs, (tuple, list)):
-            world_outputs = (world_outputs,)
-
-        pixel_outputs = wcs_out.world_to_pixel(*world_outputs)
-
-        if wcs_out.pixel_n_dim == 1:
-            pixel_outputs = (pixel_outputs,)
-
-        for ipix in range(wcs_out.pixel_n_dim):
-            if ipix in pixel_out_indices:
-                outputs[ipix] = np.broadcast_to(pixel_outputs[ipix], original_shape)
-
-    return outputs[0] if wcs_out.pixel_n_dim == 1 else outputs
+    from scipy.optimize import least_squares
+    from .wcs import Sip
+    import copy
+
+    xp, yp = xy
+    lon, lat = world_coords.data.lon.deg, world_coords.data.lat.deg
+
+    # verify input
+    if (proj_point != 'center') and (type(proj_point) != type(world_coords)):
+        raise ValueError("proj_point must be set to 'center', or an" +
+                         "`~astropy.coordinates.SkyCoord` object with " +
+                         "a pair of points.")
+    if proj_point != 'center':
+        assert proj_point.size == 1
+
+    proj_codes = [
+        'AZP', 'SZP', 'TAN', 'STG', 'SIN', 'ARC', 'ZEA', 'AIR', 'CYP',
+        'CEA', 'CAR', 'MER', 'SFL', 'PAR', 'MOL', 'AIT', 'COP', 'COE',
+        'COD', 'COO', 'BON', 'PCO', 'TSC', 'CSC', 'QSC', 'HPX', 'XPH'
+    ]
+    if type(projection) == str:
+        if projection not in proj_codes:
+            raise ValueError("Must specify valid projection code from list of " 
+                             + "supported types: ", ', '.join(proj_codes))
+        # empty wcs to fill in with fit values
+        wcs = celestial_frame_to_wcs(frame=world_coords.frame,
+                                     projection=projection)
+    else: #if projection is not string, should be wcs object. use as template.
+        wcs = copy.deepcopy(projection)
+        wcs.cdelt = (1., 1.) # make sure cdelt is 1
+        wcs.sip = None
+
+    # Change PC to CD, since cdelt will be set to 1
+    if wcs.wcs.has_pc():
+        wcs.wcs.cd = wcs.wcs.pc
+        wcs.wcs.__delattr__('pc')
+
+    if type(sip_distortion) != bool:
+       raise ValueError("sip_distortion must be set to True or False.")
+    if (sip_distortion is True) & (type(degree) != int):
+        raise ValueError("If sip_distorion is True, an integer value for " +
+                         "the polynomial order must be provided.")
+
+    # set pixel_shape to span of input points
+    wcs.pixel_shape = (max(xp)-min(xp), max(yp)-min(yp))
+
+    # determine CRVAL from input
+    close = lambda l, p: p[np.where(np.abs(l) == min(np.abs(l)))[0][0]]
+    if str(proj_point) == 'center':  # use center of input points
+        wcs.wcs.crval = ((max(lon)+min(lon))/2., (max(lat)+min(lat))/2.)
+        wcs.wcs.crpix = ((max(xp)+min(xp))/2., (max(yp)+min(yp))/2.)
```

Comment:
This will fail for a footprint that crosses the 360->0 degree crossover. In this case the CRVAL lon ends up being near 175 deg, and the wcs fitting never really recovers. This limitation should maybe be mentioned here in the doc string?

Reply:
if i can't figure out a fix for this today ill just make a note of it in the docstring. let me know if you have any ideas - i'll see if I can use something in coordinates to generalize how the center of points is computed

Reply:
I'm ok with this being a "known issue" to fix later, but if possible it would be nice to add a test case that covers this but does @xfail, and then we can make an issue in the tracker that's basically just "make the xfail not fail".

Reply:
Yeah, I couldn't come up with a good fix for it either when I encountered it. I ended up just avoiding using 'center' if I didn't know where a footprint was going to land.

Reply:
issue #5766 has a solution for this, testing it now

Reply:
maybe you could test this on one of the cases that was failing before?